docs/gc.md

```diff
@@ -110,11 +110,45 @@ A special mode of the `gencon` policy is known as *Concurrent Scavenge* (`-Xgc:c
+
+JNI critical is used when the programmer wants direct addressability of the object. To do that, JNI needs the object to be in a contiguous region of memory, or at least make it think the object is in a contiguous region. For that to happen, JNI critical creates a temporary array on the size of total elements in the original array. It then copies element by element to this temporary array. After the JNI critical is done it copies everything back, again element by element from the temporary array to the arraylet. This operation is quite expensive. Additionally, since arrays are stored in a discontiguous form, it hinders the JIT (Just in time Compile) to perform array optimizations. Double mapping arraylets was introduced to solve these problems.
+
+#### Double Mapping
+The initial idea of double mapping was to make a large array, which is divided across the heap in different chunks, look like a contiguous array.
+
+For double mapping to happen we leverage the System Virtualization. Virtual memory address space on 64-bit systems is very large, 64 bits worth in fact, compared to 32 bits in the 32-bit systems counterpart. However, physical memory is limited to a size much smaller in most cases. Using more virtual memory address space does not really cost the application anything, but using more physical memory would. Therefore, instead of using more physical memory and perform expensive copies, we map two different virtual memory addresses to the same physical memory address.
+
+![The diagram is explained in the surrounding text](./cr/arraylet_doublemap.png "Arraylet double mapping diagram")
+
+With double mapping, whenever JNI or JIT uses arraylets, they won’t need to create and/or copy element by element to a temporary array. The double mapping is created as soon as the array in the Java program is instantiated. Therefore, JNI and JIT only need to reference the contiguous memory double mapped mirror of the original array. By doing so, it saves a lot of time because the contiguous view of the array which is chucked in its arraylet leaves was already created and only needs to be referenced by JNI or JIT. Modifications made to the contiguous representation of the arraylet are reflected on the discontiguous one and vice-versa.
+
+Double mapping is enabled by default. If one wishes to disable it he must pass the command line option `-XXgc:disableArrayletDoubleMapping`.
+
+*Note:* Double mapping is only supported in Linux systems. Also, double mapping does not support systems that has huge pages as default (normally 2MB) yet. In such systems if double mapping is requested, it will get disabled. In future releases double mapping will support systems with huge pages as well as other platforms.
```

Comment:
I would change `double mapping does not support systems that has huge pages as default (normally 2MB) yet.` to something like `double mapping can not be supported if object heap happen to be allocated in Large (Huge) Pages. In such systems double mapping would be ignored even requested explicitly. Please use -Xlp:objectheap:pagesize option to configure object heap in small pages (for example -Xlp:objectheap:pagesize=4k).  Please use -verbose:sizes for reference of page sizes available in the system.`
I would remove `In such systems if double mapping is requested, it will get disabled. In future releases double mapping will support systems with huge pages as well as other platforms.`.

It can be rephrased better but point is double mapping feature is dependable from type (manifested in size) of pages object heap happen to be allocated with. An extra action might be required to make it work. Such action is request for object heap to be allocated in smallest pages using `-Xlp:objectheap:pagesize` (need hyperlink to option). Size of smallest page can be checked from `-verbose:sizes` (need hyperlink to option) output. (There is a confusing moment this output says `available large page sizes: 4K`)
I would avoid of mentioning any future improvements as well